pkg/controller/privatelink/privatelink_controller.go

```diff
@@ -0,0 +1,486 @@
+	logger.Debug("reconciling cluster deployment")
+	recobsrv := hivemetrics.NewReconcileObserver(ControllerName, logger)
+	defer recobsrv.ObserveControllerReconcileTime()
+
+	if r.hubActuator == nil {
+		logger.Error("hub actuator is not set")
+		return reconcile.Result{}, errors.New("hub actuator is not set")
+	}
+
+	// Fetch the ClusterDeployment instance
+	cd := &hivev1.ClusterDeployment{}
+	err := r.Get(context.TODO(), request.NamespacedName, cd)
+	if apierrors.IsNotFound(err) {
+		logger.Debug("cluster deployment not found")
+		return reconcile.Result{}, nil
+	}
+	if err != nil {
+		// Error reading the object - requeue the request.
+		logger.WithError(err).Error("error getting ClusterDeployment")
+		return reconcile.Result{}, err
+	}
+	logger = controllerutils.AddLogFields(controllerutils.MetaObjectLogTagger{Object: cd}, logger)
+
+	if paused, err := strconv.ParseBool(cd.Annotations[constants.ReconcilePauseAnnotation]); err == nil && paused {
+		logger.Info("skipping reconcile due to ClusterDeployment pause annotation")
+		return reconcile.Result{}, nil
+	}
+
+	// Initialize cluster deployment conditions if not present
+	newConditions, changed := conditions.InitializeConditions(cd)
+	if changed {
+		cd.Status.Conditions = newConditions
+		logger.Info("initializing private link controller conditions")
+		if err := r.Status().Update(context.TODO(), cd); err != nil {
+			logger.WithError(err).Log(controllerutils.LogLevel(err), "failed to update cluster deployment status")
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	}
+
+	var clientActuator Actuator
+	var privateLinkEnabled bool
+	switch {
+	case cd.Spec.Platform.AWS != nil:
+		if cd.Spec.Platform.AWS.PrivateLink == nil {
+			logger.Debug("controller cannot service the clusterdeployment, so skipping")
+			return reconcile.Result{}, nil
+		}
+		privateLinkEnabled = cd.Spec.Platform.AWS.PrivateLink.Enabled
+		clientActuator, err = r.GetActuator(configv1.AWSPlatformType, logger)
+		if err != nil {
+			logger.WithError(err).Log(controllerutils.LogLevel(err), "failed to get client actuator")
+			return reconcile.Result{}, err
+		}
+	default:
+		logger.Debug("controller cannot service the clusterdeployment, so skipping")
+		return reconcile.Result{}, nil
+	}
+
+	if !privateLinkEnabled {
+		if r.cleanupRequired(cd, clientActuator) {
+			// private link was disabled for this cluster so cleanup is required.
+			return r.cleanupClusterDeployment(cd, clientActuator, logger)
+		}
+
+		logger.Debug("cluster deployment does not have private link enabled, so skipping")
+		return reconcile.Result{}, nil
+	}
+
+	if cd.DeletionTimestamp != nil {
+		return r.cleanupClusterDeployment(cd, clientActuator, logger)
+	}
+
+	// Add finalizer if not already present
+	if !controllerutils.HasFinalizer(cd, finalizer) {
+		logger.Debug("adding finalizer to ClusterDeployment")
+		controllerutils.AddFinalizer(cd, finalizer)
```

Comment:
We discussed having the privatelink and awsprivatelink operators be mutually exclusive, with the active one always migrating the finalizer and conditions to itself. In other words, the privatelink operator, when active, would migrate the awsprivatelink finalizer and conditions to privatelink finalizer and conditions. At the same time, when the awsprivatelink operator is active, it would migrate the privatelink finalizer and conditions back to awsprivatelink finalizer and conditions. This would enable admins to switch between the two to test and easily revert while also making sure the migration happens.